src/V3Hasher.cpp

```diff
@@ -29,40 +31,290 @@ class HasherVisitor final : public AstNVisitor {
-                // For identical nodes, the type should be the same thus
-                // dtypep should be the same too
-                m_lowerHash = V3Hash(m_lowerHash,
-                                     V3Hash(V3Hash(nodep->type() << 6), V3Hash(nodep->dtypep())));
-                // Now update m_lowerHash for our children's (and next children) contributions
-                iterateChildrenConst(nodep);
-                // Store the hash value
-                if (m_cacheInUser4) { nodep->user4(m_lowerHash.fullValue()); }
-                thisHash = m_lowerHash;
-            }
+    inline void walk(AstNode* nodep) {  //
+        if (VL_LIKELY(nodep)) iterate(nodep);
+    }
+
+    V3Hash hashNode(AstNode* nodep, bool hashDType, bool hashChildren, std::function<void()>&& f) {
+        if (m_cacheInUser4 && nodep->user4()) {
+            return V3Hash(nodep->user4());
+        } else {
+            VL_RESTORER(m_hash);
+            // Reset accumulator
+            m_hash = V3Hash(nodep->type());  // Node type
+            f();  // Node specific hash
+            if (hashDType && nodep != nodep->dtypep()) walk(nodep->dtypep());  // Node dtype
+            if (hashChildren) iterateChildrenConst(nodep);  // Children
+            if (m_cacheInUser4) { nodep->user4(m_hash.value()); }
+            return m_hash;
         }
-        // Update what will become the above node's hash
-        m_lowerHash += m_cacheInUser4 ? V3Hash(nodep->user4()) : thisHash;
     }
 
+    // VISITORS
+
+    // Each visitor below contributes to the hash any node specific content
+    // that is not dependent on either of the following, as these are
+    // included by default by hashNode:
+    // - Node type (as given by AstNode::type())
+    // - Node dtype (unless !hashDType)
+    // - child nodes (unless !hashChildren)
+    //
+    // The hash must be stable, which means in particular it cannot rely on
+    // pointer values, or any other value that might differ between separate
+    // invocations of Verilator over the same design.
+    //
+    // Note there is a circularity problem where some child nodes can back
+    // to their ancestral nodes via member pointers, which can lead to an
+    // infinite traversal. To break this, nodes that are subject to such
+    // referencing and represent code which can reasonably be assumed not to
+    // be equivalent to any other code, are hashed either by name (e.g.:
+    // AstNodeModule), or by unique identifier (e.g.: AstNodeUOrStructDType).
+
+#define VISIT(kind, hashDType, hashChildren, body) \
+    virtual void visit(kind* nodep) override { \
+        const V3Hash hash = hashNode(nodep, hashDType, hashChildren, [=]() body); \
+        m_hash += hash; \
+    }
+
+    //------------------------------------------------------------
+    // AstNode - Warns to help find missing cases
+
+    VISIT(AstNode, true, true, {  //
+        UINFO(4, "WARNING! Hashing node as AstNode: " << nodep);
```

Comment:
```suggestion
#ifdef VL_DEBUG
            UINFO(0, "Unknown node type in V3Hasher: " << nodep->prettyTypeName() << endl);
#endif
 ``

Reply:
Also run regressions with this commented to always print and see if anything strange pops out please.

Reply:
I did run whith this on, I actually has nodep->v3fatalSrc while deveoping, but I didn't wan to leave that in as I'm sure there is some code out there that is not covered byt the testsuite, and would fail on this. With the warning the hash might just be sub-optimal. I will do the VL_DEBUG thing though, that is a good change.